[BugFix] Two problem for restore (#31790)

srlch opened the pull request:
Problem 1:
The problem between restore and tablet meta checkpoint thread still happen, some case does not cover yet.

Problem 2:
If Tablet report and restore execute concurrently, the restored replica will be set bad and lost replica

Fixes #issue

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [x] 3.0
  - [x] 2.5

wanpengfei-git commented:
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 19 / 20 (95.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TabletInvertedIndex.java|19|20|95.00%|[552]|

wanpengfei-git commented:
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 3 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/tablet_manager.cpp|0|3|00.00%|[1235, 1237, 1238]|

srlch committed:
[BugFix] Two problem for restore (#31790)

Problem 1:
The problem between restore and tablet meta checkpoint thread still happen, some case does not cover yet.

Problem 2:
If Tablet report and restore execute concurrently, the restored replica will be set bad and lost replica

Signed-off-by: srlch <linzichao@starrocks.com>

wanpengfei-git commented:
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 19 / 20 (95.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/TabletInvertedIndex.java|19|20|95.00%|[552]|

wanpengfei-git commented:
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 0 / 4 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/storage/tablet.cpp|0|2|00.00%|[1340, 1341]|
|:large_blue_circle:|src/runtime/snapshot_loader.cpp|0|1|00.00%|[733]|
|:large_blue_circle:|src/storage/tablet.h|0|1|00.00%|[317]|

wanpengfei-git commented:
@Mergifyio backport branch-3.2

wanpengfei-git commented:
@Mergifyio backport branch-3.1

wanpengfei-git commented:
@Mergifyio backport branch-3.0

wanpengfei-git commented:
@Mergifyio backport branch-2.5
